File: leetcode/stack/valid_parentheses.py
class Solution:
    def isValid(self, s: str) -> bool:
        stack = []
        close_to_open = {')': '(', '}': '{', ']': '['}

        for char in s:
            if char in close_to_open:
                if not stack or stack[-1] != close_to_open[char]:
                    return False
                stack.pop()
            else:
                stack.append(char)

        return len(stack) == 0

# Try it out — press Run
sol = Solution()
print(sol.isValid("[]"))      # True
print(sol.isValid("([{}])"))  # True
print(sol.isValid("[(])"))    # False
print(sol.isValid("(]"))      # False

File: leetcode/stack/valid_parentheses.ts
function isValid(s: string): boolean {
  const stack: string[] = [];
  const closeToOpen: Record<string, string> = {
    ')': '(',
    '}': '{',
    ']': '[',
  };

  for (const char of s) {
    if (char in closeToOpen) {
      if (stack.length === 0 || stack[stack.length - 1] !== closeToOpen[char]) {
        return false;
      }
      stack.pop();
    } else {
      stack.push(char);
    }
  }

  return stack.length === 0;
}

// Try it out — press Run
console.log(isValid("[]"));      // true
console.log(isValid("([{}])")); // true
console.log(isValid("[(])"));   // false
console.log(isValid("(]"));     // false

export {};
